earth2studio/io/async_zarr.py

```diff
@@ -0,0 +1,639 @@
+        FSSpec file system factory method. This is a callable object that should return
+        an instance of the desired filesystem to use, by default LocalFileSystem
+    blocking : bool, optional
+        Blocking write calls in the synchronous API. When set to fall, the IO backend
+        will execute write calls in seperate threads. Users should call the `close()`
+        API to ensure all threads have finished / cleaned up, by default True
+    pool_size : int, optional
+        The thread / async loop pool used with the synchronous write API in non-blocking
+        mode, by default 4
+    async_timeout : int, optional
+        Async operation timeout for a given write operation, by default 600
+    zarr_kwargs : dict[str, Any], optional
+        Additional key work arguments to provide to the ` zarr.api.asynchronous.open`
+        function, by default {"mode": "a"}
+    zarr_codecs: CompressorsLike, optional
+        Compression codec to use when creating any new arrays. Sharding is not supported
+        for thread safety at the moment. If None, will use the default compressor, by
+        default None
+    """
+
+    def __init__(
+        self,
+        file_name: str,
+        index_coords: CoordSystem = OrderedDict({}),
+        fs_factory: Callable[..., fsspec.spec.AbstractFileSystem] = LocalFileSystem,
+        blocking: bool = True,
+        pool_size: int = 4,
+        async_timeout: int = 600,
+        zarr_kwargs: dict[str, Any] = {"mode": "a"},
+        zarr_codecs: CompressorsLike = None,
+    ) -> None:
+        # May need to trigger warning about this, needed to handle multi-threading!
+        # But silent for now since people wont know what this means / get confused by an error message I think
+        AsyncFileSystem.cachable = False
+
+        try:
+            zarr_version = version("zarr")
+            zarr_major_version = int(zarr_version.split(".")[0])
+        except Exception:
+            zarr_major_version = 2
+
+        if zarr_major_version < 3:
+            raise ImportError("This IO store only support Zarr 3.0 and above")
+
+        if not callable(fs_factory):
+            raise TypeError(
+                "fs_factory must be a callable that returns a fsspec.spec.AbstractFileSystem"
+            )
+
+        self.overwrite = False  # Not formally supported
+        self.index_coords = self._scrub_coordinates(
+            index_coords
+        )  # TODO: Need to validate these somewhere, should just make a 1:1 match requirement
+        self.zarr_codecs = zarr_codecs
+
+        # Async / multi-thread items
+        self.blocking = blocking
+        if blocking:
+            pool_size = 1
+        self.async_timeout = async_timeout
+        self.io_futures: list[concurrent.futures._base.Future] = []
+        self.pool_index = 0
+        self.loop_pool = self._initialize_loop_pool(pool_size)
+        self.fs_pool = []
+        self.zarr_pool = []
+        logger.warning(
```

Comment:
nit: does this need to be a warning or could it just be debug? Would add to chatter in multi-process runs